extendr-api/src/wrapper/integers.rs

```diff
@@ -0,0 +1,101 @@
+use super::*;
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Integers {
+    pub(crate) robj: Robj,
+}
+
+impl Default for Integers {
+    fn default() -> Self {
+        Integers::new(0)
+    }
+}
+
+impl Integers {
+    /// Create a new vector of integers.
+    /// ```
+    /// use extendr_api::prelude::*;
+    /// test! {
+    ///     let vec = Integers::new(10);
+    ///     assert_eq!(vec.is_integer(), true);
+    ///     assert_eq!(vec.len(), 10);
+    /// }
+    /// ```
+    pub fn new(len: usize) -> Integers {
+        let iter = (0..len).map(|_| 0);
+        Integers::from_values(iter)
+    }
+
+    /// Wrapper for creating ALTREP integer (INTSXP) vectors from iterators.
+    /// The iterator must be exact, clonable and implement Debug.
+    ///
+    /// ```
+    /// use extendr_api::prelude::*;
+    /// test! {
+    ///     // Short (<64k) vectors are allocated.
+    ///     let vec = Integers::from_values((0..3).map(|i| 2-i));
+    ///     assert_eq!(vec.is_altrep(), false);
+    ///     assert_eq!(r!(vec.clone()), r!([2, 1, 0]));
+    ///     assert_eq!(vec.elt(1), 1);
+    ///     let mut dest = [0; 2];
+    ///     vec.get_region(1, &mut dest);
+    ///     assert_eq!(dest, [1, 0]);
+    ///
+    ///     // Long (>=64k) vectors are lazy altrep objects.
+    ///     let vec = Integers::from_values(0..1000000000);
+    ///     assert_eq!(vec.is_altrep(), true);
+    ///     assert_eq!(vec.elt(12345678), 12345678);
+    ///     let mut dest = [0; 2];
+    ///     vec.get_region(12345678, &mut dest);
+    ///     assert_eq!(dest, [12345678, 12345679]);
+    /// }
+    /// ```
+    pub fn from_values<V>(values: V) -> Self
+    where
+        V: IntoIterator,
+        V::IntoIter: ExactSizeIterator + std::fmt::Debug + Clone + 'static + std::any::Any,
+        V::Item: Into<i32>,
+    {
+        single_threaded(|| {
+            let values: V::IntoIter = values.into_iter();
+
+            let robj = if values.len() >= 65536 {
+                Altrep::make_altinteger_from_iterator(values)
+                    .try_into()
+                    .unwrap()
+            } else {
+                let mut robj = Robj::alloc_vector(INTSXP, values.len());
+                let dest: &mut [i32] = robj.as_typed_slice_mut().unwrap();
+
+                for (d, v) in dest.iter_mut().zip(values) {
+                    *d = v.into();
+                }
+                robj
+            };
+            Self { robj }
+        })
+    }
+
+    /// Get a single element from the vector.
+    pub fn elt(&self, index: usize) -> i32 {
+        unsafe { INTEGER_ELT(self.get(), index as R_xlen_t) }
+    }
```

Comment:
What I've been arguing on #261 is that this kind of method should be named as `elt_unchecked()` (and probably marked as `unsafe` function), and implement another `elt()` that does validation. We might need another variant that returns `Rint`.

Reply:
If we mark it as `unsafe` we can't have it part of the public API though, can we? (Actually, I still think we will eventually need unsafe functions in the public API for maximum usability.)

Reply:
To be clear, I think this function `elt_unchecked()` is "safe" enough to be included in the public API, in the sense that it won't destroy some memory region or panic suddenly. But, I thought it should be labelled as `unsafe` when it's users' responsibility to ensure the input is valid like this one:

https://doc.rust-lang.org/std/primitive.f32.html#method.to_int_unchecked

I'm yet to figure out the semantics of `unsafe`, so I might be wrong here.

Reply:
I was just commenting on the fact that we recently removed an unsafe public function because some linter complained, but I think a library such as extendr needs unsafe public functions. So I'm on board with marking this as unsafe.

Reply:
Ah, https://github.com/extendr/extendr/issues/54? But I didn't succeed to reproduce the warning at that time (https://github.com/extendr/extendr/issues/54#issuecomment-757086375).

This might be the one. This indicates it's fine to have `unsafe` public function as long as it has `Safety` section.

```
warning: unsafe function's docs miss `# Safety` section
  --> extendr-api/src/wrapper/integers.rs:80:5
   |
80 | /     pub unsafe fn elt(&self, index: usize) -> i32 {
81 | |         INTEGER_ELT(self.get(), index as R_xlen_t)
82 | |     }
   | |_____^
   |
   = note: `#[warn(clippy::missing_safety_doc)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_safety_doc
```

Reply:
Yes, I don't think elt will ever be efficient, so returning a validated Result always may be the best bet.

It should probably return a Result<Rint>, also.

get_region can do all its checks once for a number of elements and call "skip" on the iterator
which may take time depending on optimisation levels. It is better for performing vector operations.

We might want a get_region_repeat which repeats the vector when doing operations on different size vectors.

Reply:
I'm keen to implement parallel region operations, also.

Reply:
What's going to be the simple interface when somebody doesn't care about performance and has a smallish vector. An iterator?

Reply:
Just give it an iterator like this:

```
let x : Vec<i32> = ...
let y : Vec<i32> = ...
let plus_one = Integers::from_values(x.iter().map(|&x| x + 1);
let xplusy = Integers::from_values(x.iter().zip(y.iter())map(|(&x, &y)| x + y);

```

I'd need to check them.

Reply:
Thanks, this looks good. I was thinking about the other direction, though, e.g. add two R vectors and store the result in a rust vector.

Reply:
Yes, correct.

Reply:
The `FromIterator` is what we need here.

https://doc.rust-lang.org/std/iter/trait.FromIterator.html

Reply:
Done.